core/src/test/scala/unit/kafka/log/UnifiedLogTest.scala

```diff
@@ -3455,6 +3455,21 @@ class UnifiedLogTest {
     assertFalse(newDir.exists())
   }
 
+  @Test
+  def testMaybeUpdateHighWatermarkAsFollower(): Unit = {
+    val logConfig = LogTestUtils.createLogConfig()
+    val log = createLog(logDir, logConfig)
+
+    for (i <- 0 until 100) {
+      val records = TestUtils.singletonRecords(value = s"test$i".getBytes)
+      log.appendAsLeader(records, leaderEpoch = 0)
+    }
+
+    assertEquals(Some(100L), log.maybeUpdateHighWatermark(100L))
```

Comment:
nit: Could we add `assertEquals(Some(99L), log.maybeUpdateHighWatermark(99L))` as well?

Reply:
The sequence is a bit weird. I would have added `99L` just before `100L`. Something like this:

```
    assertEquals(Some(99L), log.maybeUpdateHighWatermark(99L))
    assertEquals(None, log.maybeUpdateHighWatermark(99L))

    assertEquals(Some(100L), log.maybeUpdateHighWatermark(100L))
    assertEquals(None, log.maybeUpdateHighWatermark(100L))

    // Bound by the log end offset.
    assertEquals(None, log.maybeUpdateHighWatermark(101L))
```